Add option to continue pipeline when Jenkins stage fail - issue #685

peterpd opened the pull request:
Issue:
https://github.com/spinnaker/spinnaker/issues/685#issuecomment-201213764

peterpd commented:
@tomas: PTAL
thanks
peter

peterpd commented:
@tomaslin : PTAL
thanks,
peter

anotherchrisberry commented:
First, sorry for the very late feedback to this pull request. We've had a number of core team members on vacation over the past couple of weeks and a few things got away from us.

I'm on board conceptually with what this is trying to accomplish, but the UX is not quite right. We really have three possible scenarios to model if the Jenkins stage fails:
 1. ignore the failure (failPipeline: false, continuePipeline: true)
 2. halt this _branch_ of the pipeline (failPipeline: false, continuePipeline: false)
 3. halt the _entire_ pipeline (failPipeline: true, continuePipeline: false)

With the two checkboxes, there's this fourth state which doesn't make sense: both checkboxes are checked.

I'd like to see this implemented as a group of radio buttons, using the options above as labels. The implementation will be a little more complicated - you'll need to manage the state of the radio buttons outside the scope of the stage object - but should be a lot clearer, something like:
```html
<div class="form-group">
  <label class="col-md-2 col-md-offset-1 sm-label-right">If job fails</label>
  <div class="col-md-6">
    <label class="radio-inline">
      <input type="radio"
                  ng-model="viewState.failureOption"
                  ng-change="failureOptionChanged()"
                  ng-value="ignore"/> ignore the failure
    </label>
    <!-- the other two options -->
  </div>
</div>
```
then handle initialization in the controller, add the method to handle the ng-change event, etc.

Also, "Failed Continue" is confusing as the status. It's a really hard status to label, it turns out - we spent twenty minutes trying to come up with something. We came up with:
 * ignore
 * disregard
 * unstable
 * failed ignoring results
 * dismiss
but none of those are great, either. So we need to find a better term for the status here. 

Last thing: the color is too close to the awaiting manual judgment color (#F0AD4E), which is going to be confusing. We should figure out a different color - maybe a yellow? I'll check with our designer on that.

I'll get an update in here once we figure out a color, and we're open to suggestions on the status name.

peterpd commented:
@anotherchrisberry : comment for UX is right on, we'll make the changes

peterpd commented:
@anotherchrisberry : made changes per your comments.  tested OK. One issue is radio buttons not aligned well, and you may have a better idea.

ajordens commented:
Bump @anotherchrisberry

anotherchrisberry commented on app/scripts/modules/core/pipeline/config/stages/jenkins/jenkinsStage.html:
```diff
@@ -79,11 +79,31 @@ <h4 class="text-left" >Job Parameters</h4>
+               ng-change="jenkinsStageCtrl.failureOptionChanged()"
+               value="stop"/> Halt this branch of the pipeline
+      </label>
+      <label class="radio-inline col-md-offset-1">
```
remove the `col-md-offset-1` class from the labels - that will fix the positioning

anotherchrisberry commented on app/scripts/modules/core/pipeline/config/stages/jenkins/jenkinsStage.html:
```diff
@@ -79,11 +79,31 @@ <h4 class="text-left" >Job Parameters</h4>
+               name="failureOption"
+               ng-model="viewState.failureOption"
+               ng-change="jenkinsStageCtrl.failureOptionChanged()"
+               value="ignore"/> Ignore the failure
```
I don't think these labels should not be capitalized - the option forms the rest of the sentence started by "If job fails", so it reads a little cleaner as "ignore the failure".

anotherchrisberry commented on app/scripts/modules/core/pipeline/config/stages/jenkins/jenkinsStage.html:
```diff
@@ -79,11 +79,31 @@ <h4 class="text-left" >Job Parameters</h4>
+               name="failureOption"
+               ng-model="viewState.failureOption"
+               ng-change="jenkinsStageCtrl.failureOptionChanged()"
+               value="fail"/> Halt the entire pipeline
```
I think we want some help bubbles here to further explain how the options differ, e.g. 
* "Immediately halts execution of all running stages and fails the entire execution"
* "Prevents any stages that depend on this stage from running, but allows other branches of the pipeline to run"
* "Continues execution of dowstream stages, marking this stages as failed/continuing"

or something like that.

anotherchrisberry commented:
I'm okay with FAILED_CONTINUE as a status, since nobody has come up with anything better. We can always change it down the line.

peterpd commented:
Changed per your comments from Chris
  added tool tips for the three radio buttons
  changed first letter of labels to lower case
  removed off-set for alignment.  -- still not ideal. 1st radio button aligned properly, but 2nd and third radio buttons shift a bit to the right

anotherchrisberry commented:
Sorry, I led you down a bad path with that markup. Try this:
```html
<div class="radio">
  <label>
    <input type="radio"
           name="failureOption"
           ng-model="viewState.failureOption"
           ng-change="jenkinsStageCtrl.failureOptionChanged()"
           value="ignore"/> ignore the failure
  </label>
</div>
```

anotherchrisberry commented on app/scripts/modules/core/pipeline/config/stages/jenkins/jenkinsStage.js:
```diff
@@ -102,6 +104,31 @@ module.exports = angular.module('spinnaker.core.pipeline.stage.jenkinsStage', [
+        initValue = 'ignore';
+      }
+      return initValue;
+    };
```
Since this method is only called once via the `ng-init` code in the template, it doesn't need to be exposed on the controller and can be rewritten as:
```javascript
function initializeFailureOption() {
  let initValue = '';
  if ($scope.stage.failPipeline === true && $scope.stage.continuePipeline === false) {
    initValue = 'fail';
  } else if ($scope.stage.failPipeline === false && $scope.stage.continuePipeline === false) {
    initValue = 'stop';
  } else if ($scope.stage.failPipeline === false && $scope.stage.continuePipeline === true) {
    initValue = 'ignore';
  }
  $scope.viewState.failureOption = initValue;
}

initializeFailureOption();
```
This keeps the controller a little cleaner, and lets you get rid of the `ng-init` attribute in the template.

peterpd commented:
that alignment works better. by now you know I'm not a UI developer.
thanks,
peter

peterpd commented:
refactored the js function for initialization.
thanks,
peter

anotherchrisberry commented:
Thanks for your persistence!
